sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java

```diff
@@ -104,18 +123,47 @@ public static Read read() {
    * matching a sharding pattern), with each element of the input collection encoded into its own
    * line.
    */
-  public static Write write() {
-    return new AutoValue_TextIO_Write.Builder()
+  public static Write<String> write() {
+    return TextIO.writeCustomType(new IdentityFormatter<String>());
+  }
+
+  /**
+   * A {@link PTransform} that writes a {@link PCollection} to a text file (or multiple text files
+   * matching a sharding pattern), with each element of the input collection encoded into its own
+   * line.
+   *
+   * <p>This version allows you to apply {@link TextIO} writes to a PCollection of a custom type
+   * {@link T}, along with a format function that converts the input type {@link T} to the String
+   * that will be written to the file. The advantage of this is it  allows a user-provided
+   * {@link DynamicDestinations} object, set via {@link Write#withDynamicDestinations} to examine
+   * the user's custom type when choosing a destination.
+   */
+  public static <T> Write<T> writeCustomType(SerializableFunction<T, String> formatFunction) {
+    return new AutoValue_TextIO_Write.Builder<T>()
         .setFilenamePrefix(null)
+        .setTempDirectory(null)
         .setShardTemplate(null)
         .setFilenameSuffix(null)
         .setFilenamePolicy(null)
+        .setDynamicDestinations(null)
+        .setFormatFunction(formatFunction)
         .setWritableByteChannelFactory(FileBasedSink.CompressionType.UNCOMPRESSED)
         .setWindowedWrites(false)
         .setNumShards(0)
         .build();
   }
 
+  /**
+   * This is a convenience builder for dynamically writing to multiple file destinations. The
+   * transform takes in {@code KV<DefaultFilenamePolicy.Params, String>}s mapping file lines to
+   * a default filename policy for writing (the params contains a prefix, suffix, and filename
+   * template). A params must also be provided to specify how to write out empty PCollections.
+   */
+  public static Write<KV<Params, String>> writeDynamic(Params emptyDestination) {
```

Comment:
This will shuffle a lot of Params, right? - I'm convinced that this helper is at all a good idea. If you get rid of it, you can also get rid of DefaultDynamicDestinations and ExtractValueFormatter I guess.